lib/commands/copy.js

```diff
@@ -0,0 +1,137 @@
+      log: this.npm.log,
+    }
+    const destination = args[0]
+    const omit = new Set(this.npm.flatOptions.omit)
+
+    const tree = await new Arborist(opts).loadActual()
+
+    // map of node to location in destination.
+    const destinations = new Map()
+
+    // calculate the root set of packages.
+    if (includeWorkspaceRoot) {
+      const to = join(destination, tree.location)
+      destinations.set(tree, to)
+    }
+    for (const edge of tree.edgesOut.values()) {
+      if (edge.workspace && workspaces.has(edge.to.realpath)) {
+        const to = join(destination, edge.to.location)
+        destinations.set(edge.to, to)
+      }
+    }
+
+    // copy the root set of packages and their dependencies.
+    const tasks = []
+    for (const [node, dest] of destinations) {
+      if (node.isLink && node.target) {
+        const targetPath = destinations.get(node.target)
+        if (targetPath == null) {
+          // This is the first time the link target was seen, it will be the
+          // only copy in dest, other links to the same target will link to
+          // this copy.
+          destinations.set(node.target, dest)
+        } else {
+          // The link target is already in the destination
+          tasks.push(relativeSymlink(targetPath, dest))
+        }
+      } else {
+        if (node.isWorkspace || node.isRoot) {
+          // workspace and root packages have not been published so they may
+          // have files that should be excluded.
+          tasks.push(copyPacklist(node.target.realpath, dest))
+        } else {
+          // copy the modules files but not dependencies.
+          const nm = join(node.realpath, 'node_modules')
+          tasks.push(fs.cp(node.realpath, dest, {
+            recursive: true,
+            errorOnExist: false,
+            filter: src => src !== nm,
+          }))
+        }
+
+        // add dependency edges to the queue.
+        for (const edge of node.edgesOut.values()) {
+          if (!omit.has(edge.type) && edge.to != null) {
+            destinations.set(
+              edge.to,
+              join(
+                destinations.get(edge.to.parent) || destination,
+                relative(edge.to.parent.location, edge.to.location)))
+          }
+        }
+      }
+    }
+    await Promise.all(tasks)
+  }
+}
+module.exports = Copy
+
+async function copyPacklist (from, to) {
+  for (const file of await packlist({ path: from })) {
+    // packlist will include bundled node_modules. ignore it because we're
+    // already handling copying dependencies.
+    if (file.startsWith('node_modules/')) {
+      continue
+    }
+
+    // using recursive copy because packlist doesn't list directories.
+    // TODO what is npm's preferred recursive copy?
+    await fs.cp(
+      join(from, file),
+      join(to, file),
+      { recursive: true, errorOnExist: false })
+  }
+}
+
+async function relativeSymlink (target, path) {
+  await fs.mkdir(dirname(path), { recursive: true })
```

Comment:
I think the mkdir `force` option should be set to true. Got an error without it on a module that was already there.

Reply:
i've spent more time on this EEXIST issue, and it seems that pushing fs.cp calls inside tasks isn't safe. mkdir operations happen concurrently under the hood multiple for a same folder, which results in this error